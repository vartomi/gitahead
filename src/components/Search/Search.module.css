.container {
    width: 700px;
    position: relative;
}

.input {
    display: block;
    margin: 0 auto;
    height: 40px;
    width: 100%;
    padding: 0 8px;
    height: 48px;
    font-size: 24px;
    border: 1px solid #ddd;
}

.input:focus {
    border: 1px solid #aaa;
    outline: none;
}

.input::placeholder {
    color: #bbb;
}