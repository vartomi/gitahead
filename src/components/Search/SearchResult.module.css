.container {
    position: absolute;
    border: 1px solid #eee;
    background-color: white;
    width: 100%;
    box-shadow: 0px 5px 10px #ccc;
    z-index: 1;
}