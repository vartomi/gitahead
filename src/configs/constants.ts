export const THRESHOLD_CHARACTER_NUMBER_FOR_SEARCH: number = 2;
export const SIMULATED_DELAY: number = 1000;

export const PLACEHOLDER_FOR_SEARCH_INPUT: string = 'Search for GitHub users...';
export const GITHUB_API_USERS: string = 'https://api.github.com/users';

export const GITHUB_QUERY_PAGE_USERS: string = 'https://github.com/search';

export enum Keys {
    Enter = 13,
    Tab = 9,
};