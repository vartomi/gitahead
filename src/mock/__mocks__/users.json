[
    {
        "login": "test-login",
        "id": 1,
        "node_id": "test",
        "avatar_url": "test-avatar",
        "gravatar_id": "",
        "url": "test",
        "html_url": "test",
        "followers_url": "test/followers",
        "following_url": "test/following{/other_user}",
        "gists_url": "test/gists{/gist_id}",
        "starred_url": "test/starred{/owner}{/repo}",
        "subscriptions_url": "test/subscriptions",
        "organizations_url": "test/orgs",
        "repos_url": "test/repos",
        "events_url": "test/events{/privacy}",
        "received_events_url": "test/received_events",
        "type": "User",
        "site_admin": false,
        "name": "Test Data",
        "company": "test",
        "blog": null,
        "location": null,
        "email": null,
        "hireable": null,
        "bio": "",
        "public_repos": 0,
        "public_gists": 0,
        "followers": 0,
        "following": 0,
        "created_at": "2009-08-01T18:39:25Z",
        "updated_at": "2020-01-27T18:17:21Z"
    }
]