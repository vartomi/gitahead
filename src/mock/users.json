[
  {
    "login": "mojombo",
    "id": 1,
    "node_id": "MDQ6VXNlcjE=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/1?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/mojombo",
    "html_url": "https://github.com/mojombo",
    "followers_url": "https://api.github.com/users/mojombo/followers",
    "following_url": "https://api.github.com/users/mojombo/following{/other_user}",
    "gists_url": "https://api.github.com/users/mojombo/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/mojombo/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mojombo/subscriptions",
    "organizations_url": "https://api.github.com/users/mojombo/orgs",
    "repos_url": "https://api.github.com/users/mojombo/repos",
    "events_url": "https://api.github.com/users/mojombo/events{/privacy}",
    "received_events_url": "https://api.github.com/users/mojombo/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Tom Preston-Werner",
    "company": null,
    "blog": "http://tom.preston-werner.com",
    "location": "San Francisco",
    "email": null,
    "hireable": null,
    "bio": null,
    "public_repos": 61,
    "public_gists": 62,
    "followers": 21770,
    "following": 11,
    "created_at": "2007-10-20T05:24:19Z",
    "updated_at": "2020-01-18T05:35:17Z"
  },
  {
    "login": "defunkt",
    "id": 2,
    "node_id": "MDQ6VXNlcjI=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/2?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/defunkt",
    "html_url": "https://github.com/defunkt",
    "followers_url": "https://api.github.com/users/defunkt/followers",
    "following_url": "https://api.github.com/users/defunkt/following{/other_user}",
    "gists_url": "https://api.github.com/users/defunkt/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/defunkt/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/defunkt/subscriptions",
    "organizations_url": "https://api.github.com/users/defunkt/orgs",
    "repos_url": "https://api.github.com/users/defunkt/repos",
    "events_url": "https://api.github.com/users/defunkt/events{/privacy}",
    "received_events_url": "https://api.github.com/users/defunkt/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Chris Wanstrath",
    "company": null,
    "blog": "http://chriswanstrath.com/",
    "location": null,
    "email": null,
    "hireable": null,
    "bio": "üçî",
    "public_repos": 107,
    "public_gists": 273,
    "followers": 20886,
    "following": 210,
    "created_at": "2007-10-20T05:24:19Z",
    "updated_at": "2019-11-01T21:56:00Z"
  },
  {
    "login": "addyosmani",
    "id": 110953,
    "node_id": "MDQ6VXNlcjExMDk1Mw==",
    "avatar_url": "https://avatars2.githubusercontent.com/u/110953?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addyosmani",
    "html_url": "https://github.com/addyosmani",
    "followers_url": "https://api.github.com/users/addyosmani/followers",
    "following_url": "https://api.github.com/users/addyosmani/following{/other_user}",
    "gists_url": "https://api.github.com/users/addyosmani/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addyosmani/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addyosmani/subscriptions",
    "organizations_url": "https://api.github.com/users/addyosmani/orgs",
    "repos_url": "https://api.github.com/users/addyosmani/repos",
    "events_url": "https://api.github.com/users/addyosmani/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addyosmani/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy Osmani",
    "company": "Google",
    "blog": "https://www.addyosmani.com",
    "location": "Mountain View, California",
    "email": null,
    "hireable": null,
    "bio": "Engineering Manager at Google working on Chrome & Web Platform",
    "public_repos": 307,
    "public_gists": 172,
    "followers": 34886,
    "following": 254,
    "created_at": "2009-08-01T18:39:25Z",
    "updated_at": "2020-01-27T18:17:21Z"
  },
  {
    "login": "addy",
    "id": 4543783,
    "node_id": "MDQ6VXNlcjQ1NDM3ODM=",
    "avatar_url": "https://avatars1.githubusercontent.com/u/4543783?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addy",
    "html_url": "https://github.com/addy",
    "followers_url": "https://api.github.com/users/addy/followers",
    "following_url": "https://api.github.com/users/addy/following{/other_user}",
    "gists_url": "https://api.github.com/users/addy/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addy/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addy/subscriptions",
    "organizations_url": "https://api.github.com/users/addy/orgs",
    "repos_url": "https://api.github.com/users/addy/repos",
    "events_url": "https://api.github.com/users/addy/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addy/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addison Shaw",
    "company": "Cerner",
    "blog": "https://addy.sh/",
    "location": "Kansas City, Missouri",
    "email": null,
    "hireable": null,
    "bio": "Software Engineer @ Cerner Corporation",
    "public_repos": 32,
    "public_gists": 13,
    "followers": 37,
    "following": 53,
    "created_at": "2013-05-27T22:57:16Z",
    "updated_at": "2020-01-18T15:27:02Z"
  },
  {
    "login": "addykim",
    "id": 7654513,
    "node_id": "MDQ6VXNlcjc2NTQ1MTM=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/7654513?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addykim",
    "html_url": "https://github.com/addykim",
    "followers_url": "https://api.github.com/users/addykim/followers",
    "following_url": "https://api.github.com/users/addykim/following{/other_user}",
    "gists_url": "https://api.github.com/users/addykim/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addykim/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addykim/subscriptions",
    "organizations_url": "https://api.github.com/users/addykim/orgs",
    "repos_url": "https://api.github.com/users/addykim/repos",
    "events_url": "https://api.github.com/users/addykim/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addykim/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy Kim",
    "company": "Ubiquity6",
    "blog": "",
    "location": "San Francisco, CA",
    "email": null,
    "hireable": true,
    "bio": null,
    "public_repos": 51,
    "public_gists": 1,
    "followers": 24,
    "following": 20,
    "created_at": "2014-05-21T06:24:55Z",
    "updated_at": "2019-12-09T18:57:13Z"
  },
  {
    "login": "nerdyaddy",
    "id": 4039883,
    "node_id": "MDQ6VXNlcjQwMzk4ODM=",
    "avatar_url": "https://avatars1.githubusercontent.com/u/4039883?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/nerdyaddy",
    "html_url": "https://github.com/nerdyaddy",
    "followers_url": "https://api.github.com/users/nerdyaddy/followers",
    "following_url": "https://api.github.com/users/nerdyaddy/following{/other_user}",
    "gists_url": "https://api.github.com/users/nerdyaddy/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/nerdyaddy/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/nerdyaddy/subscriptions",
    "organizations_url": "https://api.github.com/users/nerdyaddy/orgs",
    "repos_url": "https://api.github.com/users/nerdyaddy/repos",
    "events_url": "https://api.github.com/users/nerdyaddy/events{/privacy}",
    "received_events_url": "https://api.github.com/users/nerdyaddy/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy",
    "company": null,
    "blog": "",
    "location": "Mumbai",
    "email": null,
    "hireable": null,
    "bio": null,
    "public_repos": 4,
    "public_gists": 0,
    "followers": 3,
    "following": 0,
    "created_at": "2013-04-02T18:11:39Z",
    "updated_at": "2018-01-22T12:44:25Z"
  }
]