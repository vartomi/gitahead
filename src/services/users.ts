import mockUserList from '../mock/users.json';
import { GITHUB_API_USERS, GITHUB_QUERY_PAGE_USERS, SIMULATED_DELAY } from '../configs/constants';

export interface User {
    id: number;
    login: string;
    avatar: string;
    url: string;
    name: string;
    blog: string | null;
    bio: string | null;
    email: string | null;
    company: string | null;
    location: string | null;
    nrOfRepositories: number;
    nrOfFollowers: number;
    nrOfFollowing: number;
    joinedAt: string;
}

export const mapRawToUser: (rawData: any) => User = function ({
    id,
    avatar_url,
    login,
    name,
    bio,
    blog,
    email,
    public_repos,
    followers,
    following,
    created_at,
    html_url,
    company,
    location
}): User {
    return {
        id,
        login,
        avatar: avatar_url,
        url: html_url,
        name,
        blog,
        bio,
        email,
        location,
        company,
        nrOfRepositories: public_repos,
        nrOfFollowers: followers,
        nrOfFollowing: following,
        joinedAt: created_at
    };
}

let users: User[] = [];

export const getRawData = async function (link: string): Promise<[]> {
    const response: Response = await fetch(link);
    if (!response.ok) {
        throw new Error(response.statusText);
    }

    let rawData: [] = await response.json();
    return rawData;
}


export const getUsers = async function (withApi?: boolean): Promise<User[]> {
    if (withApi) {
        let rawData: any = await getRawData(GITHUB_API_USERS)
        rawData = await Promise.all(rawData.map(async (data: any) => {
            const meta: any = await getRawData(data.url);
            return { ...data, ...meta };
        }));

        users = rawData.map(mapRawToUser);
        return users;
    } else {
        users = mockUserList.map(mapRawToUser);

        return new Promise((resolve, _) => {
            setTimeout(() => resolve(users), SIMULATED_DELAY);
        });
    }
}


export const findUsers = function (searchValue: string): Promise<User[]> {
    const lowerCaseSearchValue = searchValue.toLowerCase();
    return new Promise((resolve, _) => {
        const result: User[] = users.filter(user =>
            (user.login && user.login.toLowerCase().startsWith(lowerCaseSearchValue)) ||
            (user.name && user.name.toLowerCase().startsWith(lowerCaseSearchValue))
        ).slice(0, 5);
        setTimeout(() => resolve(result), SIMULATED_DELAY);
    });
};

export const searchOnGithub = function (searchValue: string): void {
    window.open(`${GITHUB_QUERY_PAGE_USERS}?q=${searchValue}&type=users`);
}