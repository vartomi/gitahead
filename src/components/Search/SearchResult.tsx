import React from 'react';
import styles from './SearchResult.module.css';
import { User } from '../../services/users';
import UserCard from '../UserProfile/UserCard';

type SearchResultProps = {
    users: User[];
    onSelect: any;
}

const SearchResult: React.FC<SearchResultProps> = ({ users, onSelect }) => {
    return (
        <div className={styles.container}>
            {users.map(user => <UserCard user={user} onSelect={onSelect} />)}
        </div>
    );
}

export default SearchResult;
