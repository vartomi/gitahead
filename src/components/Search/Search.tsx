import React, { useState, useCallback } from 'react';
import styles from './Search.module.css';
import SearchResult from './SearchResult';
import { User, findUsers, searchOnGithub } from '../../services/users';
import { THRESHOLD_CHARACTER_NUMBER_FOR_SEARCH, PLACEHOLDER_FOR_SEARCH_INPUT, Keys } from '../../configs/constants';
import ProgressBar from '../ProgressBar/ProgressBar';
import debounce from 'lodash.debounce';

type SearchProps = {
    onSelect: (user: User) => void;
}

const Search: React.FC<SearchProps> = ({ onSelect }) => {
    const [searchValue, setSearchValue]: [string, any] = useState('');
    const [top5Result, setTop5Result]: [User[], any] = useState([]);
    const [isLoading, setLoading]: [boolean, any] = useState(false);

    const debouncedSearch = useCallback(debounce((value: string) => {
        if (value.length >= THRESHOLD_CHARACTER_NUMBER_FOR_SEARCH) {
            setLoading(true);
            findUsers(value).then(result => { setLoading(false); setTop5Result(result) });
        } else {
            setTop5Result([]);
        }
    }, 300), []);

    const onSearchChange = useCallback(({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
        setSearchValue(value);
        debouncedSearch(value);
    }, [debouncedSearch]);

    const onSelectUser = useCallback((user: User) => {
        setSearchValue(user.name);
        setTop5Result([]);
        onSelect(user);
    }, [onSelect]);

    const onKeyUp = useCallback(({ keyCode }: React.KeyboardEvent<HTMLInputElement>) => {
        if (keyCode === Keys.Tab && top5Result.length > 0) {
            onSelectUser(top5Result[0]);
        } else if (keyCode === Keys.Enter) {
            searchOnGithub(searchValue);
        }
    }, [onSelectUser, top5Result, searchValue]);

    return (
        <div className={styles.container} onKeyUp={onKeyUp}>
            <input
                id={'search-input'}
                className={styles.input}
                placeholder={PLACEHOLDER_FOR_SEARCH_INPUT}
                type="text"
                value={searchValue}
                onChange={onSearchChange}
            />
            {isLoading && <ProgressBar compact={true} />}
            {top5Result.length > 0 && <SearchResult users={top5Result} onSelect={onSelectUser} />}
        </div>
    );
}

export default Search;
