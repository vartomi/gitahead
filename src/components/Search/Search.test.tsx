import { shallow } from 'enzyme'
import Search from './Search';
import React from 'react';
import { Keys } from '../../configs/constants';
import { searchOnGithub } from '../../services/users';
import SearchResult from './SearchResult';

jest.mock('../../services/users', () => ({
    findUsers: () => Promise.resolve([{
        id: 1,
        login: 'login',
        avatar: 'avatar',
        url: 'url',
        name: 'name',
        blog: 'blog',
        bio: 'bio',
        email: 'email',
        company: 'company',
        location: 'location',
        nrOfRepositories: 0,
        nrOfFollowers: 0,
        nrOfFollowing: 0,
        joinedAt: 'date'
    }]),
    searchOnGithub: jest.fn(() => { })
}));

const onSelectMock = jest.fn(() => { });

describe('Search component', () => {
    it('should be rendered', () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        expect(wrapper.find('input').prop('id')).toEqual('search-input');
    });

    it('should setSearchValue', () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        wrapper.find('input').simulate('change', { target: { value: 'login' } });

        expect(wrapper.find('input').prop('value')).toEqual('login');
    });

    it('should call searchOnGithub in case of Enter key', async () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        wrapper.find('input').simulate('change', { target: { value: 'login' } });

        await new Promise(resolve => setTimeout(resolve, 500));
        wrapper.simulate('keyup', { keyCode: Keys.Enter });

        expect(searchOnGithub).toBeCalled();
    });

    it('should call onSelect in case of Tab key', async () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        wrapper.find('input').simulate('change', { target: { value: 'login' } });

        await new Promise(resolve => setTimeout(resolve, 500));
        wrapper.simulate('keyup', { keyCode: Keys.Tab });
        wrapper.update();

        expect(onSelectMock).toBeCalled();
    });

    it('should render SearchResult with user in case of matching user', async () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        wrapper.find('input').simulate('change', { target: { value: 'login' } });

        await new Promise(resolve => setTimeout(resolve, 500));
        wrapper.update();

        expect(wrapper.find(SearchResult).length).toBeTruthy();
    });

    it('should not render SearchResult with less character typed in', async () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        wrapper.find('input').simulate('change', { target: { value: 't' } });

        await new Promise(resolve => setTimeout(resolve, 500));
        wrapper.update();

        expect(wrapper.find(SearchResult).length).not.toBeTruthy();
    });

    it('should render snapshot', () => {
        const wrapper = shallow(<Search onSelect={onSelectMock} />);
        expect(wrapper).toMatchSnapshot();
    });
});