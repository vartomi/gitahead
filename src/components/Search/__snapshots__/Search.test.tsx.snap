// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render snapshot 1`] = `
<div
  className="container"
  onKeyUp={[Function]}
>
  <input
    className="input"
    id="search-input"
    onChange={[Function]}
    placeholder="Search for GitHub users..."
    type="text"
    value=""
  />
</div>
`;
